// @midwayjs/core/dist/common/serviceFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceFactory = void 0;
const extend_1 = require("../util/extend");
/**
 * 多客户端工厂实现
 */
class ServiceFactory {
    constructor() {
        this.clients = new Map();
        this.options = {};
    }
    async initClients(options = {}) {
        this.options = options;
        // merge options.client to options.clients['default']
        if (options.client) {
            options.clients = options.clients || {};
            options.clients['default'] = options.clients['default'] || {};
            (0, extend_1.extend)(true, options.clients['default'], options.client);
        }
        // multi client
        if (options.clients) {
            for (const id of Object.keys(options.clients)) {
                await this.createInstance(options.clients[id], id);
            }
        }
    }
    get(id = 'default') {
        return this.clients.get(id);
    }
    has(id) {
        return this.clients.has(id);
    }
    async createInstance(config, clientName) {
        // options.default will be merge in to options.clients[id]
        config = (0, extend_1.extend)(true, {}, this.options['default'], config);
        const client = await this.createClient(config, clientName);
        if (client) {
            if (clientName) {
                this.clients.set(clientName, client);
            }
            return client;
        }
    }
    async destroyClient(client) { }
    async stop() {
        for (const value of this.clients.values()) {
            await this.destroyClient(value);
        }
    }
    getDefaultClientName() {
        return this.options['defaultClientName'];
    }
}
exports.ServiceFactory = ServiceFactory;
