// @cool-midway/sms/ali.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmsAli = void 0;
const Core = require("@alicloud/pop-core");
const core_1 = require("@midwayjs/core");
const core_2 = require("@cool-midway/core");
/**
 * 阿里云短信
 */
let SmsAli = class SmsAli {
    /**
     * 配置
     * @param config
     */
    setConfig(config) {
        this.config = config;
    }
    /**
     * 发送短信
     * @param phone 手机号
     * @param params 参数
     * @param config  signName 签名 template 模板
     * @returns
     */
    async send(phone, params, config) {
        const { accessKeyId, accessKeySecret } = this.config;
        if (!accessKeyId || !accessKeyId) {
            throw new core_2.CoolCommException('请配置阿里云短信');
        }
        if (!config) {
            config = {
                signName: this.config.signName,
                template: this.config.template,
            };
        }
        const client = new Core({
            accessKeyId,
            accessKeySecret,
            endpoint: 'https://dysmsapi.aliyuncs.com',
            // endpoint: 'https://cs.cn-hangzhou.aliyuncs.com',
            apiVersion: '2017-05-25',
            // apiVersion: '2018-04-18',
        });
        const data = {
            RegionId: 'cn-shanghai',
            PhoneNumbers: phone,
            signName: config.signName,
            templateCode: config.template,
            TemplateParam: JSON.stringify(params),
        };
        return await client.request('SendSms', data, {
            method: 'POST',
        });
    }
};
__decorate([
    (0, core_1.Config)('cool.sms.ali'),
    __metadata("design:type", Object)
], SmsAli.prototype, "config", void 0);
SmsAli = __decorate([
    (0, core_1.Provide)()
], SmsAli);
exports.SmsAli = SmsAli;
