// @cool-midway/sms/tx.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmsTx = void 0;
const core_1 = require("@midwayjs/core");
const tencentcloud = require("tencentcloud-sdk-nodejs");
const core_2 = require("@cool-midway/core");
/**
 * 腾讯云短信
 */
let SmsTx = class SmsTx {
    /**
     * 配置
     * @param config
     */
    setConfig(config) {
        this.config = config;
    }
    /**
     * 发送短信
     * @param phone 手机号
     * @param params 参数
     * @param config  signName 签名 template 模板
     * @returns
     */
    async send(phone, params, config) {
        const { appId, secretId, secretKey } = this.config;
        if (!config) {
            config = {
                signName: this.config.signName,
                template: this.config.template,
            };
        }
        if (!appId || !secretId || !secretKey) {
            throw new core_2.CoolCommException('请配置腾讯云短信');
        }
        const smsClient = tencentcloud.sms.v20210111.Client;
        const client = new smsClient({
            credential: {
                secretId,
                secretKey,
            },
            region: 'ap-guangzhou',
            profile: {
                signMethod: 'HmacSHA256',
                httpProfile: {
                    reqMethod: 'POST',
                    reqTimeout: 30,
                    endpoint: 'sms.tencentcloudapi.com',
                },
            },
        });
        const data = {
            SmsSdkAppId: appId,
            SignName: config.signName,
            TemplateId: config.template,
            TemplateParamSet: params,
            PhoneNumberSet: [`+86${phone}`],
        };
        return client.SendSms(data);
    }
};
__decorate([
    (0, core_1.Config)('cool.sms.tx'),
    __metadata("design:type", Object)
], SmsTx.prototype, "config", void 0);
SmsTx = __decorate([
    (0, core_1.Provide)()
], SmsTx);
exports.SmsTx = SmsTx;
