// @midwayjs/locate/dist/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findCommonDir = exports.findFile = exports.findDependenciesByAST = exports.filterModule = exports.propertyExists = exports.safeGetProperty = exports.safeReadJSON = exports.exists = void 0;
const fse = require("fs-extra");
exports.exists = async (file) => {
    return fse.pathExists(file);
};
exports.safeReadJSON = async (file) => {
    if (await exports.exists(file)) {
        try {
            return fse.readJSON(file);
        }
        catch (_) {
            return {};
        }
    }
    else {
        return {};
    }
};
exports.safeGetProperty = (json, property) => {
    if (!json) {
        return null;
    }
    let properties = property;
    if (typeof property === 'string') {
        properties = property.split('.');
    }
    const currentProperty = properties.shift();
    if (properties.length > 0 && typeof json[currentProperty] === 'object') {
        return exports.safeGetProperty(json[currentProperty], properties);
    }
    return json[currentProperty];
};
exports.propertyExists = (json, properties) => {
    for (let propertyText of properties) {
        const data = exports.safeGetProperty(json, propertyText);
        if (data) {
            return true;
        }
    }
    return false;
};
const nativeModule = [
    'assert',
    'async_hooks',
    'buffer',
    'child_process',
    'cluster',
    'crypto',
    'dgram',
    'dns',
    'domain',
    'events',
    'fs',
    'http',
    'http2',
    'https',
    'inspector',
    'module',
    'net',
    'os',
    'path',
    'perf_hooks',
    'process',
    'punycode',
    'querystring',
    'readline',
    'repl',
    'stream',
    'string_decoder',
    'sys',
    'timers',
    'tls',
    'trace_events',
    'tty',
    'url',
    'util',
    'v8',
    'vm',
    'wasi',
    'worker_threads',
    'zlib',
    'config',
];
exports.filterModule = (module, modules) => {
    var _a, _b;
    // remove local module
    if (/^\./.test(module))
        return;
    // filter native module
    if (nativeModule.indexOf(module) !== -1)
        return;
    if (/^@/.test(module)) {
        // @midwayjs/abc/bbb
        if (((_a = module.match(/\//g)) === null || _a === void 0 ? void 0 : _a.length) >= 2) {
            const result = module.split('/');
            module = result[0] + '/' + result[1];
        }
    }
    else {
        // abc/bbb
        if (((_b = module.match(/\//g)) === null || _b === void 0 ? void 0 : _b.length) >= 1) {
            const result = module.split('/');
            module = result[0];
        }
    }
    modules.add(module);
};
// export const findDependencies = src => {
//   const dep = [];
//   src.replace(/(import .+ )?(from|require)\s?['"(](.+)['")]/g, (...args) => {
//     dep.push(args[3]);
//     return args[3];
//   });
//   return dep;
// };
exports.findDependenciesByAST = (source, jsx = false) => {
    const matches = [
        [/(?:^|\n|;|\s)\s*import\s+[\s\w\$\_\{\},\*\.\[\]]+\s+from\s*['"](.*?)['"]\s*;?/mg, 1],
        [/(?:^|\n|;|\s)\s*import\s*\(\s*['"](.*?)['"]\s*\)/mg, 1],
        [/(?:^|\n|;|\s)\s*import\s*['"](.*?)['"]\s*/mg, 1],
        [/(?:^|\n|;|\s)\s*import\s+[\w\$\_]+\s*from\s*['"](.*?)['"]\s*;?/mg, 1],
        [/(?:^|\n|;|\s)\s*export\s+(?:\*|\{[^\}]+\})\s+from\s+['"](.*?)['"]\s*;?/mg, 1],
        [/(?:^|\n|;|\s)\s*require\s*\(\s*['"](.*?)['"]\s*\)/mg, 1],
    ];
    const depsMap = {};
    for (const match of matches) {
        const [reg, resIndex] = match;
        let execRes;
        while (execRes = reg.exec(source)) {
            // console.log('execRes', execRes);
            depsMap[execRes[resIndex]] = true;
        }
    }
    return Object.keys(depsMap);
};
exports.findFile = async (files) => {
    for (const file of files) {
        if (await exports.exists(file)) {
            return file;
        }
    }
};
exports.findCommonDir = (files) => {
    if (files.length == 0)
        return "";
    const isWindow = files[0].indexOf('\\') != -1;
    const splitStr = isWindow ? '\\' : '/';
    const allFilePath = files.map(file => {
        const list = file.split(splitStr);
        const last = list[list.length - 1];
        if (last.indexOf('.') != -1) {
            return list.slice(0, -1);
        }
        return list;
    });
    let ans = allFilePath[0];
    for (let i = 1; i < allFilePath.length; i++) {
        let j = 0;
        for (; j < ans.length && j < allFilePath[i].length; j++) {
            if (ans[j] != allFilePath[i][j])
                break;
        }
        ans = ans.slice(0, j);
        if (!ans.length) {
            return '';
        }
    }
    return ans.join(splitStr);
};
