// @midwayjs/logger/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearAllLoggers = exports.createFileLogger = exports.createConsoleLogger = exports.createLogger = exports.loggers = exports.MidwayLoggerContainer = exports.MidwayContextLogger = exports.MidwayDelegateLogger = exports.MidwayBaseLogger = exports.displayLabels = exports.displayCommonMessage = exports.transports = exports.format = void 0;
const container_1 = require("./container");
var winston_1 = require("winston");
Object.defineProperty(exports, "format", { enumerable: true, get: function () { return winston_1.format; } });
Object.defineProperty(exports, "transports", { enumerable: true, get: function () { return winston_1.transports; } });
var format_1 = require("./format");
Object.defineProperty(exports, "displayCommonMessage", { enumerable: true, get: function () { return format_1.displayCommonMessage; } });
Object.defineProperty(exports, "displayLabels", { enumerable: true, get: function () { return format_1.displayLabels; } });
__exportStar(require("./interface"), exports);
__exportStar(require("./transport"), exports);
var logger_1 = require("./logger/logger");
Object.defineProperty(exports, "MidwayBaseLogger", { enumerable: true, get: function () { return logger_1.MidwayBaseLogger; } });
var delegate_1 = require("./logger/delegate");
Object.defineProperty(exports, "MidwayDelegateLogger", { enumerable: true, get: function () { return delegate_1.MidwayDelegateLogger; } });
var contextLogger_1 = require("./logger/contextLogger");
Object.defineProperty(exports, "MidwayContextLogger", { enumerable: true, get: function () { return contextLogger_1.MidwayContextLogger; } });
var container_2 = require("./container");
Object.defineProperty(exports, "MidwayLoggerContainer", { enumerable: true, get: function () { return container_2.MidwayLoggerContainer; } });
exports.loggers = new container_1.MidwayLoggerContainer();
const createLogger = (name, options = {}) => {
    return exports.loggers.createLogger(name, options);
};
exports.createLogger = createLogger;
const createConsoleLogger = (name, options = {}) => {
    return exports.loggers.createLogger(name, Object.assign(options, {
        disableError: true,
        disableFile: true,
    }));
};
exports.createConsoleLogger = createConsoleLogger;
const createFileLogger = (name, options = {}) => {
    return exports.loggers.createLogger(name, Object.assign(options, {
        disableConsole: true,
        disableError: true,
    }));
};
exports.createFileLogger = createFileLogger;
const clearAllLoggers = () => {
    exports.loggers.reset();
};
exports.clearAllLoggers = clearAllLoggers;
