// @midwayjs/debugger/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearDebug = exports.debugWrapper = void 0;
const child_process_1 = require("child_process");
const utils_1 = require("./utils");
__exportStar(require("./utils"), exports);
let child;
exports.debugWrapper = (options) => {
    if (options.debug || options.ts || options.child) {
        if (!child) {
            child = {
                isReady: false,
                invoke: (data) => {
                    return new Promise((resolve, reject) => {
                        const id = utils_1.getRandomId();
                        child.invokeMap[id] = { resolve, reject };
                        utils_1.sendData(child.process, { type: 'invoke', id, data });
                    });
                },
                process: null,
                invokeMap: {}
            };
            ;
            (async () => {
                const port = options.port || '9229';
                const execArgv = [];
                if (options.debug) {
                    const portIsUse = await utils_1.checkPort(port);
                    if (portIsUse) {
                        console.log('\n\n');
                        console.log(`Debug port ${port} is in use`);
                        console.log('\n\n');
                    }
                    utils_1.vscodeSupport(options);
                    execArgv.push(`--inspect=${port}`);
                }
                if (options.ts) {
                    execArgv.push('-r', 'ts-node/register');
                }
                const debugInfo = utils_1.getDebugPath();
                if (debugInfo.extensions) {
                    execArgv.push(...debugInfo.extensions);
                }
                child.process = child_process_1.fork(debugInfo.path, [
                    JSON.stringify({
                        export: options.export,
                        file: options.file,
                        port,
                        debug: options.debug
                    }),
                ], {
                    cwd: process.cwd(),
                    env: process.env,
                    execArgv
                });
                utils_1.onMessage(child.process, async (msg) => {
                    if (msg.type === 'ready') {
                        child.isReady = true;
                    }
                    else if (msg.type === 'response') {
                        const full = child.invokeMap[msg.id];
                        delete child.invokeMap[msg.id];
                        if (msg.success) {
                            if (msg.function) {
                                msg.function.forEach(functionInfo => {
                                    msg.result[functionInfo.name] = (...args) => {
                                        return new Promise((resolve, reject) => {
                                            const id = utils_1.getRandomId();
                                            child.invokeMap[id] = { resolve, reject };
                                            utils_1.sendData(child.process, {
                                                type: 'invokeInnerFunc',
                                                id,
                                                data: {
                                                    funcId: functionInfo.id,
                                                    args
                                                }
                                            });
                                        });
                                    };
                                });
                            }
                            full.resolve(msg.result);
                        }
                        else {
                            full.reject(msg.error);
                        }
                    }
                    else if (msg.type === 'childExit') {
                        exports.clearDebug(msg.exitCode);
                    }
                });
                process.on('SIGINT', () => {
                    exports.clearDebug(process.exitCode);
                });
            })();
        }
        return (...args) => {
            return waitChildReady().then((child) => {
                return child.invoke(args);
            });
        };
    }
    else {
        return utils_1.getFun(options);
    }
};
exports.clearDebug = (exitCode) => {
    if (child && child.process && child.process.pid > 0) {
        const pid = child.process.pid;
        try {
            child.process.kill(0);
            child.process.kill();
            try {
                child.process.kill(0);
                child_process_1.execSync('kill -9 ' + pid);
            }
            catch (ex) {
                void 0;
            }
        }
        catch (ex) {
            void 0;
        }
        child = null;
    }
    // 若子进程异常退出则本进程相同异常退出，以确保外层调用（jest，mocha等）能正确感知执行结果
    if (exitCode > 0) {
        process.exit(exitCode);
    }
};
// 等待子进程ready，避免在同时调用多次的时候创建多个debug子进程
const waitChildReady = () => {
    return new Promise(resolve => {
        if (child && child.isReady) {
            resolve(child);
        }
        else {
            setTimeout(() => {
                waitChildReady().then(resolve);
            }, 50);
        }
    });
};
