// @midwayjs/decorator/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerlessTriggerType = exports.MidwayFrameworkType = exports.FrameworkType = exports.MSListenerType = exports.MSProviderType = exports.InjectModeEnum = exports.ScopeEnum = exports.FORMAT = exports.FileUtils = exports.sleep = exports.Utils = exports.Types = void 0;
__exportStar(require("@midwayjs/core/dist/decorator"), exports);
var core_1 = require("@midwayjs/core");
Object.defineProperty(exports, "Types", { enumerable: true, get: function () { return core_1.Types; } });
Object.defineProperty(exports, "Utils", { enumerable: true, get: function () { return core_1.Utils; } });
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return core_1.sleep; } });
Object.defineProperty(exports, "FileUtils", { enumerable: true, get: function () { return core_1.FileUtils; } });
Object.defineProperty(exports, "FORMAT", { enumerable: true, get: function () { return core_1.FORMAT; } });
var core_2 = require("@midwayjs/core");
/**
 * @deprecated
 */
Object.defineProperty(exports, "ScopeEnum", { enumerable: true, get: function () { return core_2.ScopeEnum; } });
/**
 * @deprecated
 */
Object.defineProperty(exports, "InjectModeEnum", { enumerable: true, get: function () { return core_2.InjectModeEnum; } });
/**
 * @deprecated
 */
Object.defineProperty(exports, "MSProviderType", { enumerable: true, get: function () { return core_2.MSProviderType; } });
/**
 * @deprecated
 */
Object.defineProperty(exports, "MSListenerType", { enumerable: true, get: function () { return core_2.MSListenerType; } });
/**
 * @deprecated
 */
Object.defineProperty(exports, "FrameworkType", { enumerable: true, get: function () { return core_2.FrameworkType; } });
/**
 * @deprecated
 */
Object.defineProperty(exports, "MidwayFrameworkType", { enumerable: true, get: function () { return core_2.MidwayFrameworkType; } });
/**
 * @deprecated
 */
Object.defineProperty(exports, "ServerlessTriggerType", { enumerable: true, get: function () { return core_2.ServerlessTriggerType; } });
