// @midwayjs/core/dist/config/config.default.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util/");
const path_1 = require("path");
const constants_1 = require("../constants");
exports.default = (appInfo) => {
    var _a;
    const isDevelopment = (0, util_1.isDevelopmentEnvironment)((0, util_1.getCurrentEnvironment)());
    const logRoot = (_a = process.env[constants_1.MIDWAY_LOGGER_WRITEABLE_DIR]) !== null && _a !== void 0 ? _a : appInfo.root;
    return {
        asyncContextManager: {
            enable: false,
        },
        midwayLogger: {
            default: {
                dir: (0, path_1.join)(logRoot, 'logs', appInfo.name),
                level: 'info',
                consoleLevel: isDevelopment ? 'info' : 'warn',
                auditFileDir: '.audit',
            },
            clients: {
                coreLogger: {
                    level: isDevelopment ? 'info' : 'warn',
                    fileLogName: 'midway-core.log',
                },
                appLogger: {
                    fileLogName: 'midway-app.log',
                    aliasName: 'logger',
                },
            },
        },
        debug: {
            recordConfigMergeOrder: isDevelopment,
        },
    };
};
