// @midwayjs/view-ejs/dist/configuration.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewEJSConfiguration = void 0;
const core_1 = require("@midwayjs/core");
const View = require("@midwayjs/view");
const DefaultConfig = require("./config/config.default");
const LocalConfig = require("./config/config.local");
const view_1 = require("./view");
let ViewEJSConfiguration = class ViewEJSConfiguration {
    async onReady() {
        this.viewManager.use('ejs', view_1.EjsView);
    }
};
__decorate([
    (0, core_1.Inject)(),
    __metadata("design:type", View.ViewManager)
], ViewEJSConfiguration.prototype, "viewManager", void 0);
ViewEJSConfiguration = __decorate([
    (0, core_1.Configuration)({
        namespace: 'view-ejs',
        importConfigs: [
            {
                default: DefaultConfig,
                local: LocalConfig,
            },
        ],
        imports: [View],
    })
], ViewEJSConfiguration);
exports.ViewEJSConfiguration = ViewEJSConfiguration;
