// @midwayjs/logger/dist/logger/child.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MidwayChildLogger = void 0;
const contextLogger_1 = require("./contextLogger");
class MidwayChildLogger {
    constructor(parentLogger, options) {
        this.parentLogger = parentLogger;
        this.options = options;
    }
    getConsoleLevel() {
        return this.parentLogger.getConsoleLevel();
    }
    getFileLevel() {
        return this.parentLogger.getFileLevel();
    }
    write(...args) {
        return this.parentLogger.write(...args);
    }
    debug(...args) {
        this.transformLog('debug', args);
    }
    info(...args) {
        this.transformLog('info', args);
    }
    warn(...args) {
        this.transformLog('warn', args);
    }
    error(...args) {
        this.transformLog('error', args);
    }
    getParentLogger() {
        return this.parentLogger;
    }
    getLoggerOptions() {
        return this.options;
    }
    createContextLogger(ctx, options = {}) {
        return new contextLogger_1.MidwayContextLogger(ctx, this, {
            ...this.getLoggerOptions(),
            ...options,
        });
    }
    transformLog(level, args) {
        return this.parentLogger[level].apply(this.parentLogger, [
            ...args,
            {
                format: this.options.format,
                jsonFormat: this.options.jsonFormat,
            },
        ]);
    }
}
exports.MidwayChildLogger = MidwayChildLogger;
