// @cool-midway/core/decorator/tag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoolTag = exports.CoolUrlTag = exports.TagTypes = exports.COOL_METHOD_TAG_KEY = exports.COOL_URL_TAG_KEY = void 0;
const decorator_1 = require("@midwayjs/decorator");
exports.COOL_URL_TAG_KEY = "decorator:cool:url:tag";
exports.COOL_METHOD_TAG_KEY = "decorator:cool:method:tag";
var TagTypes;
(function (TagTypes) {
    TagTypes["IGNORE_TOKEN"] = "ignoreToken";
    TagTypes["IGNORE_SIGN"] = "ignoreSign";
})(TagTypes = exports.TagTypes || (exports.TagTypes = {}));
/**
 * 打标记
 * @param data
 * @returns
 */
function CoolUrlTag(data) {
    return (target) => {
        // 将装饰的类，绑定到该装饰器，用于后续能获取到 class
        (0, decorator_1.saveModule)(exports.COOL_URL_TAG_KEY, target);
        // 保存一些元数据信息，任意你希望存的东西
        (0, decorator_1.saveClassMetadata)(exports.COOL_URL_TAG_KEY, data, target);
    };
}
exports.CoolUrlTag = CoolUrlTag;
/**
 * 方法打标记
 * @param data
 * @returns
 */
function CoolTag(tag) {
    return (target, key, descriptor) => {
        (0, decorator_1.savePropertyDataToClass)(exports.COOL_METHOD_TAG_KEY, {
            key,
            tag
        }, target, key);
        return descriptor;
    };
}
exports.CoolTag = CoolTag;
